util/src/main/java/io/kubernetes/client/informer/cache/ReflectorRunnable.java

```diff
@@ -88,24 +89,30 @@ public void run() {
       }
       while (true) {
         if (!isActive.get()) {
-          if (watch != null) {
-            watch.close();
-            return;
-          }
+          closeWatch();
+          return;
         }
 
         try {
           if (log.isDebugEnabled()) {
             log.debug(
                 "{}#Start watch with resource version {}", apiTypeClass, lastSyncResourceVersion);
           }
-          watch =
+          Watchable<ApiType> newWatch =
               listerWatcher.watch(
                   new CallGeneratorParams(
                       Boolean.TRUE,
                       lastSyncResourceVersion,
                       Long.valueOf(Duration.ofMinutes(5).toMillis()).intValue()));
-          watchHandler(watch);
+
+          synchronized (this) {
```

Comment:
What is this synchronization protecting? I see other access to `isActive` that isn't synchronized (for example)

Reply:
@brendandburns as of now synchronized is used to protect the [watch variable](https://github.com/iyashu/java/blob/5f377e6a009682e98f57e21cafd1490c998c3905/util/src/main/java/io/kubernetes/client/informer/cache/ReflectorRunnable.java#L39) holding the current watchable instance. Essentially we want to close the watch from caller thread and re-assign it(as we re-create watches) in runnable thread thus requiring the access synchronization for it. `isActive` other hand, is atomicBoolean tracking if watch is required to be active or not (essentially from the caller thread), so that the runnable loop can return if not required.